src/hooks.rs

```diff
@@ -0,0 +1,264 @@
+unsafe extern "C" fn plrust_process_utility_hook(
+    pstmt: *mut pg_sys::PlannedStmt,
+    query_string: *const ::std::os::raw::c_char,
+    context: pg_sys::ProcessUtilityContext,
+    params: pg_sys::ParamListInfo,
+    query_env: *mut pg_sys::QueryEnvironment,
+    dest: *mut pg_sys::DestReceiver,
+    qc: *mut pg_sys::QueryCompletion,
+) {
+    plrust_process_utility_hook_internal(
+        pstmt,
+        query_string,
+        false,
+        context,
+        params,
+        query_env,
+        dest,
+        qc,
+    )
+}
+
+#[cfg(not(feature = "pg13"))]
+#[pg_guard]
+unsafe extern "C" fn plrust_process_utility_hook(
+    pstmt: *mut pg_sys::PlannedStmt,
+    query_string: *const ::std::os::raw::c_char,
+    read_only_tree: bool,
+    context: pg_sys::ProcessUtilityContext,
+    params: pg_sys::ParamListInfo,
+    query_env: *mut pg_sys::QueryEnvironment,
+    dest: *mut pg_sys::DestReceiver,
+    qc: *mut pg_sys::QueryCompletion,
+) {
+    plrust_process_utility_hook_internal(
+        pstmt,
+        query_string,
+        read_only_tree,
+        context,
+        params,
+        query_env,
+        dest,
+        qc,
+    )
+}
+
+#[allow(unused_variables)] // for `read_only_tree` under pg13
+unsafe fn plrust_process_utility_hook_internal(
+    pstmt: *mut pg_sys::PlannedStmt,
+    query_string: *const ::std::os::raw::c_char,
+    read_only_tree: bool,
+    context: pg_sys::ProcessUtilityContext,
+    params: pg_sys::ParamListInfo,
+    query_env: *mut pg_sys::QueryEnvironment,
+    dest: *mut pg_sys::DestReceiver,
+    qc: *mut pg_sys::QueryCompletion,
+) {
+    let pstmt = PgBox::from_pg(pstmt);
+    let utility_stmt = PgBox::from_pg(pstmt.utilityStmt);
+
+    // examine the UtilityStatement itself.  We're interested in three different commands,
+    // "DROP FUNCTION", "DROP SCHEMA", and "ALTER FUNCTION".  Two different node types are used
+    // for these -- `DropStmt` and `AlterFunctionStmt`
+
+    if utility_stmt.type_ == pg_sys::NodeTag_T_DropStmt {
+        let drop_stmt = PgBox::from_pg(pstmt.utilityStmt.cast::<pg_sys::DropStmt>());
+
+        // in the case of DropStmt, if the object being dropped is a FUNCTION or a SCHEMA, we'll
+        // go off and handle those two cases
+        match drop_stmt.removeType {
+            pg_sys::ObjectType_OBJECT_FUNCTION => handle_drop_function(&drop_stmt),
+            pg_sys::ObjectType_OBJECT_SCHEMA => handle_drop_schema(&drop_stmt),
+            _ => {
+                // we don't do anything for the other objects
+            }
+        }
+    } else if utility_stmt.type_ == pg_sys::NodeTag_T_AlterFunctionStmt {
+        let alter_stmt = PgBox::from_pg(pstmt.utilityStmt.cast::<pg_sys::AlterFunctionStmt>());
+
+        // and for AlterFunctioStmt, we'll just go do it.
+        handle_alter_function(&alter_stmt);
+    }
+
+    if PREVIOUS_PROCESS_UTILITY_HOOK.is_some() {
+        // previous hook must go last as we need the catalog entries this utility statement might
```

Comment:
Is your concern here that other hooks might maul the catalog entries we need here?

Reply:
I suppose you could argue we ought to go last in stead of first, in case some other extension tries to change the "STRICT" property of a plrust function.  But...

The whole hook system is a farce anyways, as changing the listed order in `shared_preload_libraries` will change what the "previous hook" is.  And it requires that extension to even bother to call the previous hook, which isn't enforced in any way.

Reply:
Thinking about this, I think we want to go before the previous hook for DROP FUNCTION/SCHEMA, but go after the hook for ALTER FUNCTION.

That seems like the right thing to do.  I'll rejigger the code.

Reply:
No I think you're right. The API isn't the greatest. We can do "best effort" but since we're dependent on ordering and what the previous hook decided to do it's all up in the air